### DISCUSSION TITLE
`lua filter` syntax for `gsub`

### DISCUSSION THREAD

**Discussion by mingsu** [quarto-dev/quarto-cli#3146] [Q&A]
I'm new to `lua filter`, kindly ask for help for simple `gsub` usage.

In `example.qmd`:

````qmd
---
title: "Example"
format: html
filters:
  - test
---

# Working without space

{{helloworld}}

with `test.lua`, expected to be AAA helloworld BBBB, working.

# Not-working with space

{{hello world}}

with `test.lua`, expected to be AAA hello world BBBB, but doesn't work.

````

In `test.lua`

````lua
return {
	{
		Str = function(elem)
			if elem.text:match("{{.*}}") then
				return elem.text:gsub("{{(.*)}}", "AAA %1 BBBB")
			else
				return elem
			end
		end,
	},
}

````

**Comment by dragonstyle** [comment]
The issue is that the space is included in the AST for Pandoc (I used `quarto preview example.qmd --to native` to look at the raw AST), so the second appearance of this looks like:

```Para [ Str "{{hello" , Space , Str "world}}" ]```

in the AST. A couple of ideas:

- You could operate on the block level (e.g. get paragraphs and walk their contents).
- These look a lot like shortcodes, have you considered simply creating a new shortcode to accomplish this? https://quarto.org/docs/extensions/shortcodes.html

In other that need to perform tasks against text ranges, I've used spans to accomplish this- so syntactically a little less terse but has other advantages. For example user could write:

```[hello world]{.replace}```

And then you can target `spans` in your filter and looks at the class names to decide what to do.

**Reply by mingsu** [reply]
Thanks!

> In other that need to perform tasks against text ranges, I've used spans to accomplish this- so syntactically a little less terse but has  other advantages. For example user could write:
> 
> [hello world]{.replace}
> 
> And then you can target spans in your filter and looks at the class names to decide what to do.


This is just what I want to do. 

More specifically, I want to change `[hello world]{.replace arguments=ARG}` to `\replace{ARG}{hello world}` in `LaTeX`.

I have tried this extension [quarto-ext/latex-environment](https://github.com/quarto-ext/latex-environment).

It can parse:

`[hello world]{.replace options=OPT}`

to 
`\replace[OPT]{hello world}`;

**But it can not parse:**


`[hello world]{.replace arguments=ARG}`

to 

`\replace{ARG}{hello world}`;

**Reply by mcanouil** [reply]
Because the named argument `arguments` is not defined the way you think.
See https://github.com/quarto-ext/latex-environment/blob/main/_extensions/latex-environment/latex-environment.lua#L80-L83

`arguments` and `options` are two different parameters.
